docs: migration guide from v3 to v4 (#4552)
vuejs/vue-cli: docs/migrating-from-v3/README.md

```

And then see the following section for detailed breaking changes introduced in each package.

------

## One-By-One Manual Migration

If you want to migrate manually and gradually, here are the tips:

### The Global `@vue/cli`

#### [Redesigned](https://github.com/vuejs/vue-cli/pull/4090) `vue upgrade`

- Before: `vue upgrade [patch | minor | major]`, and it does nothing more than installing the latest versions of Vue CLI plugins.
- After: `vue upgrade [plugin-name]`. Aside from upgrading the plugins, it can run migrators from plugins to help you automate the migration process. For more options for this command, please run `vue upgrade --help`.

#### Extra Confirmation Step To Avoid Overwriting

When running `vue invoke` / `vue add` / `vue upgrade`, there's now an [extra confirmation step](https://github.com/vuejs/vue-cli/pull/4275) if you have uncommitted changes in the current repository.

![image](https://user-images.githubusercontent.com/3277634/65588457-23db5a80-dfba-11e9-9899-9dd72efc111e.png)

#### Vue Router and Vuex Now Have Corresponding CLI Plugins

When running `vue add vuex` or `vue add router`:

- In v3, only `vuex` or `vue-router` will be added to the project;

> - In v4, there will also be `@vue/cli-plugin-vuex` & `@vue/cli-plugin-router` installed.

Reviewer: ```suggestion
- In v4, there will also be `@vue/cli-plugin-vuex` or `@vue/cli-plugin-router` installed.
```

> This currently does not make an actual difference for end-users, but such design allows us to add more features for vuex & vue-router users later.

Reviewer: Let's be consistent about naming Vuex and Vue Router 😉 
```suggestion
This currently does not make an actual difference for end-users, but such design allows us to add more features for Vuex and Vue Router users later.
```

For preset and plugin authors, there are several noteworthy changes in the two plugins:

- The default directory structure was changed:
  - `src/store.js` moved to `src/store/index.js`;
  - `src/router.js` renamed to `src/router/index.js`;

> - The `router` & `routerHistoryMode` options in `preset.json` is still supported for compatibility reason. But it's now recommended to use `plugins: { '@vue/cli-plugin-router': { historyMode: true } }` for better consistency.

Reviewer: ```suggestion
- The `router` & `routerHistoryMode` options in `preset.json` are still supported for compatibility reasons. But it's now recommended to use `plugins: { '@vue/cli-plugin-router': { historyMode: true } }` for better consistency.
```

In the past, when running the `build` command in the `development` mode, the `dist` folder layout would be different from the `production` mode. Now with the following two changes, the directory structures across all modes would be the same (file names are still different - no hashes in `development` mode):

- [#4323](https://github.com/vuejs/vue-cli/pull/4323) ensure consistent directory structure for all modes
- [#4302](https://github.com/vuejs/vue-cli/pull/4302) move dev configs into serve command

#### For SASS/SCSS Users

Previously in Vue CLI v3, we shipped with `sass-loader@7` by default.

Recently `sass-loader@8` has been out and has changed its configuration format quite a lot. Here's the release notes: <https://github.com/webpack-contrib/sass-loader/releases/tag/v8.0.0>

`@vue/cli-service` will continue to support `sass-loader@7` in v4, but we strongly recommend you to take a look at the v8 release and upgrade to the latest version.

#### For Less Users

`less-loader` v4 is incompatible with `less` >= v3.10, see <https://github.com/less/less.js/issues/3414>.
It's strongly recommended to upgrade to `less-loader@5` if your project depends on it.

#### For CSS Module Users

> - [Deprecate `css.modules` in favor of `css.requireModuleExtension`](https://github.com/vuejs/vue-cli/pull/4387). This is because we've upgraded to css-loader v3 and the config format has been changed. For a more detailed explanation please follow the link.

Reviewer: ```suggestion
- [Deprecate `css.modules` in favor of `css.requireModuleExtension`](https://github.com/vuejs/vue-cli/pull/4387). This is because we've upgraded to `css-loader` v3 and the config format has been changed. For a more detailed explanation please follow the link.
```
